p2p/src/net/mock/request_manager.rs

```diff
@@ -0,0 +1,227 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Request-response manager
+//!
+//! The request-response manager is responsible for handling all activity related to inbound/outbound
+//! requests/responses. When a new peer joins, it allocates a 16-bit wide request ID zone for it which
+//! it uses for the requests it sends to that particular peer.
+//!
+//! For inbound requests, the original request ID is stored inside the manager's storage and an ephemeral
+//! request ID is allocated for the request which is then forwarded to the frontend. This is to allow the
+//! remote peers to use whatever request IDs they want for book keeping while still being able to associate
+//! outbound responses with correct inbound requests.
+
+use crate::{
+    error::{P2pError, PeerError},
+    message,
+    net::mock::types,
+};
+use std::collections::{hash_map::Entry, HashMap, HashSet};
+
+#[derive(Debug, Default)]
+struct PeerContext {
+    ephemerals: HashSet<types::MockRequestId>,
+    request_id: u16,
+    base: u64,
+}
+
+impl PeerContext {
+    pub fn new(base: u64) -> Self {
+        Self {
+            base,
+            request_id: 0u16,
+            ephemerals: Default::default(),
+        }
+    }
+
+    pub fn allocate_request_id(&mut self) -> types::MockRequestId {
+        let id = self.base + self.request_id as u64;
+        self.request_id = self.request_id.overflowing_add(1).0;
+        types::MockRequestId::new(id)
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct RequestManager {
+    /// Registered peers
+    peers: HashMap<types::MockPeerId, PeerContext>,
+
+    /// Pending outbound requests (TODO: timeouts)
+    _pending: HashMap<types::MockRequestId, types::MockRequestId>,
+
+    /// Request ID zone for allocating unique IDs for peers
+    req_id_zone: u64,
+
+    /// Next ephemeral request ID
+    next_ephemeral: types::MockRequestId,
+
+    /// Ephemeral requests IDs which are mapped to remote peer ID/request ID pair
+    ephemeral: HashMap<types::MockRequestId, (types::MockPeerId, types::MockRequestId)>,
+}
+
+impl RequestManager {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Register peer to the request manager
+    ///
+    /// Initialize peer context and allocate request ID slice for the peer
+    pub fn register_peer(&mut self, peer_id: types::MockPeerId) -> crate::Result<()> {
+        match self.peers.entry(peer_id) {
+            Entry::Occupied(_) => Err(P2pError::PeerError(PeerError::PeerAlreadyExists)),
+            Entry::Vacant(entry) => {
+                entry.insert(PeerContext::new(self.req_id_zone));
+                self.req_id_zone =
+                    self.req_id_zone.checked_add(0x10000).expect("number of peers to be lower");
```

Comment:
Since these IDs are u64 (correct me if I'm wrong), how about having 2<spu>32</sup> zones with 2<sup>32</sup> IDs each to support the occasional odd test that needs many requests. Also is the possibility of one peers request IDs overflowing into the next zone checked somewhere so it's detected when it happens?

Reply:
I'm not entirely opposed to changing the allocation scheme to what you're proposing but I find it more likely that we'll have a test that needs more than `232` zones rather than a test that has `2^32` pending requests. The request ID's are most likely going to require some work either when request timeouts are implemented. I'd prefer doing that work later as it's not in the same rush as pubsub or fixing connectivity issues.

I'm not sure if it's possible for a peer's IDs to overflow into another peer's zone since the size of the zone is capped at `u16::MAX` and zones starts at a multiple of `0x10000`. There is currently one test checking that the allocation/overflow doesn't break at least for the normal case.

Reply:
That's fair